### DISCUSSION TITLE
Example of having different types of edges

### DISCUSSION THREAD

**Discussion by tehranixyz** [danielegrattarola/spektral#104] [Q&A] [answered]
Is there any example of applying a GNN model on a graph that has different types of edges?
Or maybe for doing this, we need to add the type of edge as the edge attribute? It would be really great if more information or samples about matrix E could be provided.

I'm experimenting with applying GNN models in order to find similar graphs. 
Daniel, do you have any recommendations on how I can achieve this goal? So for each graph, I have A, X, E matrices. 
E includes the type of edge. Given 2 graphs to a GNN model, I want to measure how similar are the vector representations of the two graphs and during training, I want to increase the similarity score of the graphs if two graphs are labeled the same otherwise, decrease the similarity score.

**Comment by danielegrattarola** [comment, marked as answer]
Hi, 

You are correct, edge types are interpreted as edge attributes in Spektral. You can have a look at the examples on molecules [here](https://github.com/danielegrattarola/spektral/blob/master/examples/graph_prediction/qm9_batch.py) for a template to start. 

For your problem, one possible way would be to have a graph embedding function like so:

```python
conv = EdgeConditionedConv(32, activation='relu')
pool = GlobalSumPool()

def embed(x, a, e): 
    x = conv([x, a, e])
    z = pool(x)

    return z[0]
```
and then: 

```python
# Dummy values
a = np.ones((10, 10))
x = np.ones((10, 3))
e = np.ones((10 * 10, 4))
graph_1 = (x, a, e)
graph_2 = (x, a, e)

# Compute similarity
z1 = embed(*graph_1)
z2 = embed(*graph_2)
sim = tf.reduce_sum(z1 * z2)
```

Hope this helps. 
Cheers

**Comment by tehranixyz** [comment]
Thank you for your useful reply.
If the features of edges depend on the type of edge, what is this the proper way to present this information?
For example, we have two types of edges (T1 and T2).
T1 edges have features named K1, K2
T2 edges have features named L1, L2  (we don't have K1 and K2 as features for T2 edges)

**Comment by danielegrattarola** [comment]
That is a setting of heterogeneous edge attributes which is not yet supported in Spektral. 
Unless K1, K2, and L1, L2 have the same size, you'll have to code your own layer. 
If they do have the same size, then a possible way of using the attributes is to simply concatenate them to the "type" feature one-hot encoded: 

```
T1, K1, K2 -> [1, 0] + [K1, K2]
T2, L1, L2 -> [0, 1] + [L1, L2]
```
